### PR TITLE
Canonical URL

### PR DESCRIPTION
**Pull request by naz**
refs #10593

- [x] add fields to posts table in the database
- [x]  allow editing through Admin API
- [x]  serve the field from Content API /posts/* endpoints
- [x]  make sure it is accessible through existing handlebars helper
- [x] make sure it is not served with API v0.1
- [x] relative/absolute url transformation, and save relative in the db if canonical url belongs to current instance

@kirrg001 tested these migration/changes on a larger dataset and didn't experience noticeable speed degradation. The remaining question here I think is: what kinds of formats do we allow for the `canonical` field? Should it support relative url's as well?

### CODE REVIEW

**Comment by gargol** [comment]
After thinking a bit about the format of the `canonical` field. It would make sense to have similar behavior like in other places: accept both relative and absolute URL's, transform it into relative if the URL is from the current instance and then serve it as absolute always?

File: core/server/data/migrations/versions/2.17/1-canonical-post.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
+const Promise = require('bluebird'),
```
Comment (kirrg001) [inline core/server/data/migrations/versions/2.17/1-canonical-post.js]: > 1-canonical-post.js

Depends which PR get's merged earlier, see https://github.com/TryGhost/Ghost/pull/10585

File: core/server/data/schema/schema.js [hunk 2]
```diff
@@ -56,7 +56,8 @@ module.exports = {
         twitter_image: {type: 'string', maxlength: 2000, nullable: true},
         twitter_title: {type: 'string', maxlength: 300, nullable: true},
         twitter_description: {type: 'string', maxlength: 500, nullable: true},
-        custom_template: {type: 'string', maxlength: 100, nullable: true}
+        custom_template: {type: 'string', maxlength: 100, nullable: true},
+        canonical: {type: 'text', maxlength: 2000, nullable: true, validations: {isEmptyOrURL: true}}
```
Comment (kirrg001) [inline core/server/data/schema/schema.js]: Was the naming discussed? IMO the more specific the better e.g. `canonical_url`.

Comment (gargol) [reply core/server/data/schema/schema.js]: Good catch. Will check!

File: core/server/data/migrations/versions/2.17/1-canonical-post.js (hunk 1 above)
Comment (gargol) [reply core/server/data/migrations/versions/2.17/1-canonical-post.js]: Will make this one `2-*` as `settings` one looks closer to land in master :)

**Comment by gargol** [comment]
@kirrg001 updated with handling of absolute/relative URLs and suggested renames :+1: did some testing, but will do a bit more tomorrow before release if we decide to push it out. Feel free to any needed changes :)

File: core/server/api/v2/utils/serializers/input/utils/url.js [hunk 3]
```diff
@@ -1,13 +1,26 @@
 const _ = require('lodash');
 const {absoluteToRelative, getBlogUrl, STATIC_IMAGE_URL_PREFIX} = require('../../../../../../services/url/utils');
 
+const handleCanonicalUrl = (url) => {
+    const blogDomain = getBlogUrl().replace(/^http(s?):\/\//, '').replace(/\/$/, '');
+    const absolute = url.replace(/^http(s?):\/\//, '');
+
+    if (absolute.startsWith(blogDomain)) {
+        return absoluteToRelative(url, {withoutSubdirectory: true});
```
Comment (kirrg001) [inline core/server/api/v2/utils/serializers/input/utils/url.js]: I have the feeling we have the check for being an external url in a couple of places in Ghost now.

File: core/server/data/migrations/versions/2.17/2-canonical-post.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
+const Promise = require('bluebird'),
```
Comment (kirrg001) [inline core/server/data/migrations/versions/2.17/2-canonical-post.js]: > 2-canonical-post.js

add-canonical-url-column.js?

The string lands in the migration database.

Not super important though. Just noticed :)

File: core/server/api/v2/utils/serializers/input/utils/url.js (hunk 3 above)
Comment (kirrg001) [reply core/server/api/v2/utils/serializers/input/utils/url.js]: Would be cool to introduce: `urlService.utils.isExternalUrl(...)`
Def doesn't have to be part of the PR :)

Comment (kirrg001) [reply core/server/api/v2/utils/serializers/input/utils/url.js]: > return absoluteToRelative(url, {withoutSubdirectory: true});

References #8221. We should handle absolute -> relative on model layer, not API layer.
But the PR can stay as is for now 👍

**Review by kirrg001 — CHANGES_REQUESTED** [review]
v0.1 serves canonical_url.

**Comment by kirrg001** [comment]
Otherwise PR works 👍 No emoji commit for now, because UI is still missing.

File: core/server/api/v2/utils/serializers/input/utils/url.js (hunk 3 above)
Comment (gargol) [reply core/server/api/v2/utils/serializers/input/utils/url.js]: > We should handle absolute -> relative on model layer, not API layer.
Yes, absolutely. `absolute_url` will be included as part of the audit for https://github.com/TryGhost/Ghost/issues/10477 :ok_hand:

File: core/server/data/migrations/versions/2.17/2-canonical-post.js (hunk 4 above)
Comment (gargol) [reply core/server/data/migrations/versions/2.17/2-canonical-post.js]: Yes, renaming it sounds good :+1: Was thinking about it for a bit, but followed the same convention as in `1-og-twitter-post.js`. Renamed, so we follow better convention in the future :)

**Comment by gargol** [comment]
> v0.1 serves canonical_url.

@kirrg001 this is the first case where we are adding a field to API v2 and don't want to serve is from API v01 (also something to consider for future v3 as well). Will check on the best way to achieve this, but looks like there are 2 options at the moment:
1. Check in model's `toJSON` method if the call is coming from API (and remove fields depending which one it is). This method is the one I'm going for but we definitely need to discuss how to deal with these types of changes and make it more dynamic.
2. Filter it out on v0.1 controller level (don't think this is a good idea or sustainable with more fields to come)

Another thing that we should probably improve, is making response format checks for both APIs based on whitelist of properties instead of blacklisting the ones that are in `schema` definition.

File: core/server/api/v2/utils/serializers/input/utils/url.js (hunk 3 above)
Comment (gargol) [reply core/server/api/v2/utils/serializers/input/utils/url.js]: > I have the feeling we have the check for being an external url in a couple of places in Ghost now.

Checked it today. This check is only performed in this input serializer, and don't think like there's enough repetition for each case to extract this specific utility function. If it needs to be used elsewhere in the codebase, then would probably make sense to extract. Leaving as is for now :+1:

Comment (kirrg001) [reply core/server/api/v2/utils/serializers/input/utils/url.js]: 🤔 search for "external" in Ghost

e.g. core/server/services/url/utils.js
e.g. core/server/helpers/img_url.js

**Pull request merged**